setendings.py

```diff
@@ -0,0 +1,169 @@
+
+def set_end_col_offset(node, last_child=None):
+    """Set the end_col_offset property by the as_string attribute.
+    Uses the end column location of the last line, instead of the widest line.
+    Need to do a custom postorder traversal to set end_col_offset property of 
+    child nodes first, since it is sometimes unset in the last child.
+    Postorder setting of children results in the assumption: if a node has the
+    propert set, then all of its children are set.
+    """
+    if hasattr(node, 'end_lineno') and hasattr(node, 'end_col_offset'):
+        if node.end_lineno and node.end_col_offset:
+            return  # reduces runtime since properties already set.
+
+    # Check: contradiction found. Revise NO_CHILDREN_TYPE list for correctness.
+    assert not last_child or type(node) not in NO_CHILDREN_TYPE, '''ERROR:️ {} 
+        node in NO_CHILDREN_TYPE has children ({}). Suggest: remove it from the 
+        list.Context:\n{}'''.format(node, last_child, node.as_string())
+
+    if hasattr(last_child, 'end_col_offset'):  # Set by last child
+        assert last_child.end_col_offset is not None, '''ERROR:️ last child 
+            end_col_offset should not be None if used to set others. 
+            Node {}.'''.format(last_child)
+        node.end_col_offset = last_child.end_col_offset
+    # Set by string of node or last_child. May not have child.
+    else:
+        # Recall: children set first in postorder traversal.
+        # Should this always set by string, is there another way?
+        set_end_col_offset_by_string(node, last_child)
+
+def _get_last_child(node):
+    """Returns the last child node, or None.
+    """
+    if node.last_child():
+        # logging.debug('\n node: {}. last_child: {}'.format(node, node.last_child()))
+        return node.last_child()
+    else:
+        # Some nodes' last_child() attribute not set, e.g. astroid.Arguments
+        # Get the first child, from the generator, and use its col_offset
+        skip_to_last_child = None  # save a reference to last child for later.
+        for skip_to_last_child in node.get_children():  # generator.
+            pass  # skip to last
+        return skip_to_last_child  # node, or None.
+
+# Define the functions to transform the nodes.
+# These functions are registered to their respective node type with the
+# register_transform function.
+# Mutates the nodes, adding the properties: `end_lineno`, `end_col_offset`.
+def set_general(node):
+    """General function called by many of the other transform functions.
+    Populate ending locations for astroid node."""
+    last_child = _get_last_child(node)
+    set_end_lineno(node, last_child)
+    set_end_col_offset(node, last_child)
+
+def set_without_children(node):
+    """Populate ending locations for nodes that are guaranteed to never have
+    children. E.g. Const.
+    """
+    last_child = _get_last_child(node)
+    set_end_lineno(node, last_child)
+    set_end_col_offset_by_string(node, last_child)
```

Comment:
Based on my comment above, you should change the signature of `set_end_col_offset_by_string` to only take in the node; last_child should always be `None` here.